Provena PR (Minor) - Feat 1512 - Generate Report

parth-kulkarni1 opened the pull request:
# Provena PR: Minor

## Checklist

-   [ ] If tests are required for this change, are they implemented?
-   [ ] Are user documentation changes required, if so, is there a task to track it and/or is it completed?
-   [ ] If migrations are required, is the process documented below?
-   [ ] If developer/system documentation updates are required, is there a task to track it and/or is it completed?
-   [ ] At least one developer has reviewed this change (unless PR is being used to mark a commit point without need for review)?
-   [ ] If change requires update to the [manual testing runsheet](https://confluence.csiro.au/pages/viewpage.action?pageId=1701138616), is there a task to track it and/or is it completed?

## Description
This feature implements the Generate Report Feature originating from a Study or Model Run Entity. 
The feature traverses from a model run and finds all associated inputs, outputs and model runs involved.

## Migrations Required
None

## Notes for reviewer
None

## Feature Branch Information

https://f1512.dev.rrap-is.com/

Deployed components:

[landing-portal](https://f1512.dev.rrap-is.com/)

[job-api](https://f1512-job-api.dev.rrap-is.com/)

[data-store-api](https://f1512-data-api.dev.rrap-is.com/)

[data-store-ui](https://f1512-data.dev.rrap-is.com/)

[auth-api](https://f1512-auth-api.dev.rrap-is.com/)

[prov-api](https://f1512-prov-api.dev.rrap-is.com/)

[prov-ui](https://f1512-prov.dev.rrap-is.com/)

[registry-api](https://f1512-registry-api.dev.rrap-is.com/)

[registry-ui](https://f1512-registry.dev.rrap-is.com/)

To tear down feature deployment: ./fb destroy 1512 export-prov-graph

PeterBaker0 commented on registry-ui/src/hooks/useGenerateReport.tsx:
```diff
@@ -0,0 +1,82 @@
+
+    // Return the values generated in the hook. 
+
+    return {
+        dataReady, 
+        submit: dataReady ? submit : undefined,
+        response: dataReady
+            ?  {
+                data: generateReportQuery.data, 
+                loading: generateReportQuery.isLoading, 
+                error: generateReportQuery.isError, 
+                success: generateReportQuery.isSuccess, 
+                fetching: generateReportQuery.isFetching,
+                errorMessage: generateReportQuery.error 
+                ? (generateReportQuery.error as string)
+                : undefined
+               }
+            
+            :undefined
+    }
```
This is a situation where you can just let typescript annotate the return types for you

parth-kulkarni1 replied:
Okay, I have updated it to return the whole object.

PeterBaker0 commented on registry-ui/src/hooks/useGenerateReportDialog.tsx:
```diff
@@ -0,0 +1,136 @@
+export const useGenerateReportDialog = (props: GenerateReportProps) => {
+
+    const [popUpOpen, setPopupOpen] = useState<boolean>(false);
+    const [depth, setDepth] = useState<string>("");
```
If you need it to be a string for your query string, then handle this later - don't muddy up the client implementation so that it can be sent easier to the API.

PeterBaker0 commented on registry-ui/src/hooks/useGenerateReportDialog.tsx:
```diff
@@ -0,0 +1,136 @@
+            <DialogContent>
+                <Stack spacing={2} direction="column">
+
+                    Select a depth for the upstream.
```
Wording here is a bit weird - but we can fine tune no stress

PeterBaker0 commented on registry-ui/src/pages/Item.tsx:
```diff
@@ -457,6 +473,7 @@ const RecordView = observer((props: {}) => {
       {versionControls.render()}
       {revertControls.render()}
       {studyLinkDialog.render()}
+      {renderedDialog}
```
all the others are functions - I'm slightly concerned that having it be a constant might be weird - but see how it goes 🤷

PeterBaker0 reviewed (commented):
Looking great - good job on commenting code well it was easy to understand what was going on. Front end is coming along well - it is a complex code base to get stuck into so good job getting some progress in there. We can fine tune wording, and check behaviour matches what we expect. 

Longer term, I suspect that when you analyse upstream/downstream you might need to keep some trace of the connection/structure - but I think your implementation fits the spec so is a good start.

Rosspet commented on prov-api/helpers/generate_report_helpers.py:
```diff
@@ -0,0 +1,223 @@
+        If the fetched entity is of type SeededItem this exception is raised.
+    """
+    
+    validation_mapping: Dict[ItemSubType, Callable] = {
+        ItemSubType.MODEL_RUN: validate_model_run_id,
+        ItemSubType.STUDY: validate_study_id
+    }
+
+    validator_func = validation_mapping.get(item_subtype)
+
+    if not validator_func:
+        raise HTTPException(status_code=400, detail="Unsupported item subtype for validation.")
```
Nice i like this approach. 

I think the error message could be enhanced to be something like "Unsupported item subtype for validation. Received subtype X. Supported types are [y,z]". You could get the supported types dynamically from the keys in the validation mapping dict defined above aswell. Its a bit longer message but can help with debugging.

parth-kulkarni1 replied:
Changed - Thanks Ross!

Rosspet commented on prov-api/helpers/generate_report_helpers.py:
```diff
@@ -0,0 +1,223 @@
+
+    node_list_parsed: List[Node] = []
+
+    if len(node_list) == 0: 
+        # Do I handle this? Can there be a case where a model run does not have any assoicated inputs/outputs?? - I don't think so.
+        raise HTTPException(status_code=400, detail="Cannot process the report generation.")
```
I also like the idea of raising errors even if its supposedly unlikely to occur - yet errors still happen. I think doing this is called [negative space programming](https://double-trouble.dev/post/negativ-space-programming/) and I get around it. 

You can also use assert statements instead like 
```assert len(node_list)!=0, "Expected non 0 node list length, cannot process report generation"``` 

Is a 400 the correct status code as well? Technically our system is designed for model runs to always have a connection so the assumption is that our system would have stuffed up in this case and resulted in a MR without connectinos so 500 would be more appropriate here. Note that if an assertion statement fails the default is 500 so you should only use them for initiating exceptions that are not a users fault (like in this scenario). 

Also, are nodes in the node_list here any connection to a model run like an organisation or study?, and not necessarily just inputs/outputs?

parth-kulkarni1 replied:
> I also like the idea of raising errors even if its supposedly unlikely to occur - yet errors still happen. I think doing this is called [negative space programming](https://double-trouble.dev/post/negativ-space-programming/) and I get around it.
> 
> You can also use assert statements instead which can initiate an exception though ideally they are wrapped in a try/except in a higher scope: `assert len(node_list)!=0, "Expected non 0 node list length, cannot process report generation"`
> 
> Is a 400 the correct status code as well? Technically our system is designed for model runs to always have a connection so the assumption is that our system would have stuffed up in this case and resulted in a MR without connectinos so 500 would be more appropriate here. Note that if an assertion statement fails and is not caught the api by default sends a 500 so you should only use them for initiating exceptions that are not a users fault (like in this scenario). Or catch the error in a high scope and raise a HTTPException with the correct status code.
> 
> Also, are nodes in the node_list here any connection to a model run like an organisation or study?, and not necessarily just inputs/outputs?

Agreed it's like a defensive programming approach. Yes, your correct it should send back a 500 status code. 

Yes they can be potential agents/create activities, but I will filter them out.

Rosspet commented on prov-api/helpers/generate_report_helpers.py:
```diff
@@ -0,0 +1,223 @@
+    if len(node_list) == 0: 
+        # Do I handle this? Can there be a case where a model run does not have any assoicated inputs/outputs?? - I don't think so.
+        raise HTTPException(status_code=400, detail="Cannot process the report generation.")
+    try:
+        for node in node_list: 
+            node_list_parsed.append(Node(**node))
+    except ValidationError as e: 
+        raise HTTPException(status_code=400, detail="Something has gone wrong in parsing this node.")
```
and also "this node" is not specific. Maybe include "... in parsing node of ID {node.id}

parth-kulkarni1 replied:
Okay I have structured the try and catch block to capture more specific details of the node. Probably not the best idea to do node.id - since it has not been parsed yet, however I did node.get('id', 'undefined id') to be safe.

Rosspet commented on prov-api/helpers/generate_report_helpers.py:
```diff
@@ -0,0 +1,223 @@
+async def generate_report_helper(starting_id: str, upstream_depth: int, shared_responses: Dict[str, List[Node]], config: Config, downstream_depth: int = 1) -> None:
+    """Generates the report by querying upstream and downstream data, parsing nodes, and filtering results.
+
+    This helper function performs upstream and downstream queries, parses the node responses, filters them 
+    for relevant data, and stores the results into a shared dictionary for further report generation.
+
+    Parameters
+    ----------
+    starting_id : str
+        The starting node ID to perform the queries from.
+    upstream_depth : int
+        The depth of the upstream query to traverse upto.
+    shared_responses : Dict[str, List[Node]]
+        A dictionary which contains the parsed and filtered nodes. 
+        Contains 'inputs', 'model_runs', and 'outputs' as keys.
+    config : Config
+        A config object containing information about the different endpoints of the system.
+    downstream_depth : int, optional
+        The depth of the downstream query to traverse to, by default 1
+    """
```
nice documentation 📝

Rosspet commented on prov-api/routes/explore/explore.py:
```diff
@@ -316,3 +321,115 @@ async def effected_agents(
+    )
+
+    # Validate the depth.
+    if depth < 1 or depth > 3: 
+        raise HTTPException(status_code=400, detail="Invalid input depth provided. Input depth is only from 1 to 3")
```
see https://docs.pydantic.dev/latest/concepts/fields/#numeric-constraints

parth-kulkarni1 replied:
Thanks for that suggestion Peter and sharing that validation trick Ross. I have changed to a POST request and integrated that field validation.

Rosspet commented on prov-api/routes/explore/explore.py:
```diff
@@ -316,3 +321,115 @@ async def effected_agents(
+
+    elif item_subtype == ItemSubType.STUDY: 
+
+        """
+        Upstream (Outputs):
+        Example: From a study, it will fetch all model runs associated with the study and their inputs, up to 3 layers upstream.
+
+        Downstream (Outputs): 
+        Example: From a study, it will fetch all relevant datasets (outputs) from the model runs linked to the study, up to 1 layer downstream.
+        """
+    
+        # Query upstream to get all linked model runs at depth 1. 
+        # This may contain more than one model run.
+        upstream_model_run_response = await explore_downstream(starting_id=node_id, depth=1, config=config)
+        assert upstream_model_run_response.graph, "Node collections not found!"
+
+        model_run_nodes: List[Node]= parse_nodes(upstream_model_run_response.graph.get('nodes', []))
+
+        # Now branch out with the model runs in here and explore them as above (depth 1-3 upstream and depth 1 downstream)
+        # The model run will be the new updated starting id.
+
+        for model_run_node in model_run_nodes:
+            if model_run_node.item_subtype == ItemSubType.MODEL_RUN:
+                model_run_id = model_run_node.id
+                # Extract the upstream and downstream entities from this node. 
+                await generate_report_helper(starting_id=model_run_id, upstream_depth=depth, shared_responses=all_filtered_responses, config=config)                
```
could you chuck this into a function?

parth-kulkarni1 replied:
I have broken this down into different sub-functions, should help with readability.

parth-kulkarni1 commented on prov-api/helpers/generate_report_helpers.py:
```diff
@@ -0,0 +1,223 @@
+
+    Returns
+    -------
+    Dict[str,List[Node]]
+        A dictionary containing the "inputs", "outputs" and "model-runs" involved within the study/model-run.
```
I have integrated a data class. I agree, a better approach and was able to add a method into it to add nodes accordingly based on their subtypes / fetched direction (upstream,downstream)

parth-kulkarni1 commented on registry-ui/src/hooks/useGenerateReport.tsx:
```diff
@@ -0,0 +1,82 @@
+
+    // Setting up the useQuery
+    const generateReportQuery = useQuery<GenerateReportResponse>({
+        queryKey: ['depth-fetch'],
```
Okay, yep thought useQuery was the best for this use-case since it was a GET request as we discussed, but since its POST now and we are making some small changes on the server side, I will change it to a UseMutate.

parth-kulkarni1 commented:
Hi @PeterBaker0, 

I have integrated your and Ross's comments from the earlier draft version. There have been some changes in the front-end patterns: 

- I have removed the React Query call-backs such as OnSuccess or OnError that was previously there. My reason for this the hooks that have been developed are not being used by other functionalities and no side effects are being done to the parent component, hence there is no real need of having or passing the callbacks from child -> parent. The state and side effects are managed within the child itself.

- Changed from useQuery to useMutate, and as you suggested instead of selectively returning attributes from the query function, I have returned the entire object. This was a good approach, as I could manage conditionally rendering error and success messages. 

- Shifted the front-end code from the Registry UI to the Shared-Libs as it will be used in Registry UI and Prov Store UI.

-  Creation of a new "Menu Item" in the Side Panel that encapsulates "View in Registry" and "Generate Report" as we discussed. 

- Added more try-catch exceptions within the back-end helper methods for robustness and better error handling, broke down the functions into more helper functions and completed doc-strings.

Sorry for the choas in merging from main to feature branch. Might be better of viewing in a code-editor itself.
